File: BlendHub/Helpers/BlenderVersionFetcher.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace BlendHub.Helpers
{
    public class BlenderVersionFetcher
    {
        private static readonly HttpClient _httpClient = new();
        private const string GitHubRawUrl = "https://raw.githubusercontent.com/DesignLipsx/blendhub/master/BlendHub/blender_versions_web.json";
        private const int TimeoutSeconds = 10;

        public BlenderVersionFetcher()
        {
            _httpClient.Timeout = TimeSpan.FromSeconds(TimeoutSeconds);
        }

        /// <summary>
        /// Asynchronously fetches Blender versions JSON from GitHub.
        /// </summary>
        /// <returns>JSON string with version data, or null if fetch fails</returns>
        public async Task<string?> FetchVersionsJsonAsync()
        {
            try
            {
                var response = await _httpClient.GetAsync(GitHubRawUrl);
                response.EnsureSuccessStatusCode();
                return await response.Content.ReadAsStringAsync();
            }
            catch (HttpRequestException ex)
            {
                System.Diagnostics.Debug.WriteLine($"HTTP Error fetching versions: {ex.Message}");
                return null;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error fetching versions: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Asynchronously fetches and deserializes Blender versions from GitHub.
        /// </summary>
        /// <typeparam name="T">Type to deserialize into</typeparam>
        /// <returns>Deserialized object or null if fetch fails</returns>
        public async Task<T?> FetchVersionsAsync<T>() where T : class
        {
            try
            {
                var json = await FetchVersionsJsonAsync();
                if (string.IsNullOrEmpty(json))
                    return null;

                return JsonSerializer.Deserialize<T>(json);
            }
            catch (JsonException ex)
            {
                System.Diagnostics.Debug.WriteLine($"JSON Error: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Fetch with fallback to local data if GitHub fetch fails.
        /// </summary>
        /// <param name="fallbackJson">Local JSON fallback data</param>
        /// <returns>JSON string (from GitHub or fallback)</returns>
        public async Task<string?> FetchVersionsWithFallbackAsync(string? fallbackJson)
        {
            var result = await FetchVersionsJsonAsync();
            return result ?? fallbackJson;
        }
    }
}

File: BlendHub/Helpers/BlenderVersionFetcher.py
"""
Utility for fetching Blender versions JSON from GitHub repository.
This can be used in your C# WinUI application via Python interop or as reference.
"""

import httpx
import asyncio
from typing import Optional, Dict, Any


class BlenderVersionFetcher:
    """Fetches Blender versions from GitHub raw content."""
    
    # GitHub raw content URL
    GITHUB_RAW_URL = "https://raw.githubusercontent.com/DesignLipsx/blendhub/master/BlendHub/blender_versions_web.json"
    TIMEOUT = 10.0
    
    @staticmethod
    async def fetch_versions() -> Optional[Dict[str, Any]]:
        """
        Asynchronously fetch Blender versions JSON from GitHub.
        
        Returns:
            Dictionary with version data or None if fetch fails
        """
        try:
            async with httpx.AsyncClient(timeout=BlenderVersionFetcher.TIMEOUT) as client:
                response = await client.get(BlenderVersionFetcher.GITHUB_RAW_URL)
                response.raise_for_status()
                return response.json()
        except httpx.HTTPError as e:
            print(f"HTTP Error: {e}")
            return None
        except Exception as e:
            print(f"Error fetching versions: {e}")
            return None
    
    @staticmethod
    async def fetch_versions_with_fallback(fallback_data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """
        Fetch versions with fallback to local data if fetch fails.
        
        Args:
            fallback_data: Local data to use if GitHub fetch fails
            
        Returns:
            Dictionary with version data (from GitHub or fallback)
        """
        versions = await BlenderVersionFetcher.fetch_versions()
        return versions if versions is not None else fallback_data or {}


# Example usage
async def main():
    fetcher = BlenderVersionFetcher()
    versions = await fetcher.fetch_versions()
    
    if versions:
        print("Successfully fetched versions:")
        print(f"Found {len(versions)} entries")
    else:
        print("Failed to fetch versions")


if __name__ == "__main__":
    asyncio.run(main())
